File: Fundamentals_of_Software_Design/C_language/Quick_sort/quick_sort.c
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "quick_sort.h"

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

size_t partition(int* arr, size_t low, size_t high) {
    size_t pivot_idx = low + rand() % (high - low + 1);
    swap(&arr[pivot_idx], &arr[high]);
    
    int pivot = arr[high];
    size_t i = low;
    
    for (size_t j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            swap(&arr[i], &arr[j]);
            i++;
        }
    }
    swap(&arr[i], &arr[high]);
    return i;
}

void quick_sort_helper(int* arr, size_t low, size_t high) {
    if (low < high) {
        size_t pi = partition(arr, low, high);
        if (pi > 0) quick_sort_helper(arr, low, pi - 1);
        quick_sort_helper(arr, pi + 1, high);
    }
}

int* quick_sort(const int* arr, size_t size) {
    int* arr_copy = (int*)malloc(size * sizeof(int));
    if (arr_copy == NULL) return NULL;
    memcpy(arr_copy, arr, size * sizeof(int));

    srand(time(NULL));
    
    if (size > 1) {
        quick_sort_helper(arr_copy, 0, size - 1);
    }
    
    return arr_copy;
}

File: Fundamentals_of_Software_Design/Python_language/Quick_sort/quick_sort.py
import random

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    
    def partition(low, high, arr):
        pivot_idx = random.randint(low, high)
        arr[pivot_idx], arr[high] = arr[high], arr[pivot_idx]
        
        pivot = arr[high]
        i = low - 1
        
        for j in range(low, high):
            if arr[j] <= pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1
    
    def quick_sort_helper(low, high, arr):
        if low < high:
            pi = partition(low, high, arr)
            quick_sort_helper(low, pi - 1, arr)
            quick_sort_helper(pi + 1, high, arr)
    
    arr_copy = arr.copy()
    quick_sort_helper(0, len(arr_copy) - 1, arr_copy)
    return arr_copy
